File: editor/en/[752]Open the Lock.py
# leetcode submit region begin(Prohibit modification and deletion)
from collections import deque


class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        dead_set = set(deadends)

        if "0000" in dead_set:
            return -1

        queue = deque(["0000"])
        dead_set.add("0000")
        count = 0

        while queue:
            size = len(queue)

            for _ in range(size):
                curr = queue.popleft()

                if curr == target:
                    return count

                next_states = self.get_next_states(curr)

                for state in next_states:
                    if state not in dead_set:
                        dead_set.add(state)
                        queue.append(state)

            count += 1

        return -1

    def get_next_states(self, curr: str) -> List[str]:
        next_states = []
        chars = list(curr)

        for i in range(4):
            curr_char = chars[i]
            chars[i] = '0' if curr_char == '9' else str(int(curr_char) + 1)
            next_states.append(''.join(chars));
            chars[i] = '9' if curr_char == '0' else str(int(curr_char) - 1)
            next_states.append(''.join(chars));
            chars[i] = curr_char

        return next_states
# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[752]Open the Lock.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function (deadends, target) {
    const deadSet = new Set(deadends);

    if (deadSet.has("0000")) {
        return -1;
    }

    const queue = ["0000"];
    deadSet.add("0000");
    let count = 0;

    while (queue.length > 0) {
        let size = queue.length;

        for (let i = 0; i < size; i++) {
            const curr = queue.shift();

            if (curr === target) {
                return count;
            }

            const nextStates = getNextStates(curr);

            for (const state of nextStates) {
                if (!deadSet.has(state)) {
                    deadSet.add(state);
                    queue.push(state);
                }
            }
        }

        count++;
    }

    return -1;
};

/**
 * Generates the next states for a given state.
 *
 * @param {string} curr - The current state.
 * @return {Array<string>} - The next states.
 */
function getNextStates(curr) {
    const nextStates = [];
    const chars = [...curr];

    for (let i = 0; i < 4; i++) {
        const currChar = chars[i];
        chars[i] = currChar === '9' ? '0' : String.fromCharCode(currChar.charCodeAt(0) + 1);
        nextStates.push(chars.join(''));
        chars[i] = currChar === '0' ? '9' : String.fromCharCode(currChar.charCodeAt(0) - 1);
        nextStates.push(chars.join(''));
        chars[i] = currChar;
    }

    return nextStates;
}

//leetcode submit region end(Prohibit modification and deletion)
